File: internal/adapter/sqlite/repo_server.go
package sqlite

import (
	"context"
	"database/sql"
	"errors"
	"fmt"
	"time"

	"github.com/jaenster/hoardarr/internal/domain/server"
)

// ServerRepo implements domain/server.Repository against SQLite.
type ServerRepo struct {
	db *DB
}

// Compile-time check.
var _ server.Repository = (*ServerRepo)(nil)

// NewServerRepo wires the repository over db.
func NewServerRepo(db *DB) *ServerRepo {
	return &ServerRepo{db: db}
}

// Save inserts a new row when ID == 0 (and back-fills the id via
// SetID), or updates an existing row otherwise.
//
// The serialization is tx-aware: if ctx carries an active *sql.Tx it
// is used; otherwise a single-statement auto-commit runs.
func (r *ServerRepo) Save(ctx context.Context, s *server.UsenetServer) error {
	if s.ID() == 0 {
		return r.insert(ctx, s)
	}
	return r.update(ctx, s)
}

func (r *ServerRepo) insert(ctx context.Context, s *server.UsenetServer) error {
	res, err := r.db.ExecCtx(ctx, `
		INSERT INTO servers(
			name, host, port, tls, username, password,
			max_conns, priority, enabled,
			backup, billing_mode, quota_bytes, used_bytes, bandwidth_bytes_per_sec,
			added_at, updated_at
		) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
	`,
		s.Name(), s.Host(), s.Port(), boolToInt(s.TLS()), nullableString(s.Username()), nullableString(s.Password()),
		s.MaxConns(), s.Priority(), boolToInt(s.Enabled()),
		boolToInt(s.Backup()), string(s.BillingMode()), s.QuotaBytes(), s.UsedBytes(), s.BandwidthBytesPerSec(),
		s.AddedAt().UnixMilli(), s.UpdatedAt().UnixMilli(),
	)
	if err != nil {
		return fmt.Errorf("insert server: %w", err)
	}
	id, err := res.LastInsertId()
	if err != nil {
		return fmt.Errorf("last insert id: %w", err)
	}
	s.SetID(server.ServerID(id))
	return nil
}

func (r *ServerRepo) update(ctx context.Context, s *server.UsenetServer) error {
	res, err := r.db.ExecCtx(ctx, `
		UPDATE servers SET
			name = ?, host = ?, port = ?, tls = ?, username = ?, password = ?,
			max_conns = ?, priority = ?, enabled = ?,
			backup = ?, billing_mode = ?, quota_bytes = ?, used_bytes = ?, bandwidth_bytes_per_sec = ?,
			updated_at = ?
		WHERE id = ?
	`,
		s.Name(), s.Host(), s.Port(), boolToInt(s.TLS()), nullableString(s.Username()), nullableString(s.Password()),
		s.MaxConns(), s.Priority(), boolToInt(s.Enabled()),
		boolToInt(s.Backup()), string(s.BillingMode()), s.QuotaBytes(), s.UsedBytes(), s.BandwidthBytesPerSec(),
		s.UpdatedAt().UnixMilli(),
		int64(s.ID()),
	)
	if err != nil {
		return fmt.Errorf("update server: %w", err)
	}
	n, err := res.RowsAffected()
	if err != nil {
		return fmt.Errorf("rows affected: %w", err)
	}
	if n == 0 {
		return server.ErrNotFound
	}
	return nil
}

// ByID returns the server with the given id, or ErrNotFound.
func (r *ServerRepo) ByID(ctx context.Context, id server.ServerID) (*server.UsenetServer, error) {
	row := r.db.QueryRowCtx(ctx, selectServerByID, int64(id))
	return scanServer(row)
}

// ByName returns the server with the given (unique) display name.
func (r *ServerRepo) ByName(ctx context.Context, name string) (*server.UsenetServer, error) {
	row := r.db.QueryRowCtx(ctx, selectServerByName, name)
	return scanServer(row)
}

// List returns all servers ordered by priority ascending then id.
func (r *ServerRepo) List(ctx context.Context) ([]*server.UsenetServer, error) {
	return r.queryServers(ctx, selectAllServers)
}

// ListEnabled returns enabled servers ordered by priority ascending,
// id ascending. Used by the download orchestrator's dispatch.
func (r *ServerRepo) ListEnabled(ctx context.Context) ([]*server.UsenetServer, error) {
	return r.queryServers(ctx, selectEnabledServers)
}

// IncrementUsedBytes bumps used_bytes by n in a single UPDATE statement.
// Used by the byte-accounting flusher in the download package; avoids
// the read-modify-write cost of going through Save() for every flush
// tick (and the lost-update risk if two flushers raced — they don't
// today but the SQL atomicity is the right shape regardless).
//
// Bumping updated_at too keeps "last seen" semantics consistent with
// every other server-state mutation.
func (r *ServerRepo) IncrementUsedBytes(ctx context.Context, id server.ServerID, n int64) error {
	if n <= 0 {
		return nil
	}
	res, err := r.db.ExecCtx(ctx, `
		UPDATE servers SET used_bytes = used_bytes + ?, updated_at = ?
		WHERE id = ?
	`, n, time.Now().UTC().UnixMilli(), int64(id))
	if err != nil {
		return fmt.Errorf("increment used_bytes: %w", err)
	}
	rows, _ := res.RowsAffected()
	if rows == 0 {
		return server.ErrNotFound
	}
	return nil
}

// Delete removes the row by id. Returns ErrNotFound if id does not
// exist.
func (r *ServerRepo) Delete(ctx context.Context, id server.ServerID) error {
	res, err := r.db.ExecCtx(ctx, `DELETE FROM servers WHERE id = ?`, int64(id))
	if err != nil {
		return fmt.Errorf("delete server: %w", err)
	}
	n, err := res.RowsAffected()
	if err != nil {
		return fmt.Errorf("rows affected: %w", err)
	}
	if n == 0 {
		return server.ErrNotFound
	}
	return nil
}

func (r *ServerRepo) queryServers(ctx context.Context, query string, args ...any) ([]*server.UsenetServer, error) {
	rows, err := r.db.QueryCtx(ctx, query, args...)
	if err != nil {
		return nil, fmt.Errorf("query servers: %w", err)
	}
	defer rows.Close()
	var out []*server.UsenetServer
	for rows.Next() {
		s, err := scanServerFromRows(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, s)
	}
	return out, rows.Err()
}

const serverColumns = `id, name, host, port, tls, username, password,
	max_conns, priority, enabled,
	backup, billing_mode, quota_bytes, used_bytes, bandwidth_bytes_per_sec,
	added_at, updated_at`

const selectServerByID = `SELECT ` + serverColumns + ` FROM servers WHERE id = ?`
const selectServerByName = `SELECT ` + serverColumns + ` FROM servers WHERE name = ?`
const selectAllServers = `SELECT ` + serverColumns + ` FROM servers ORDER BY priority ASC, id ASC`
const selectEnabledServers = `SELECT ` + serverColumns + ` FROM servers WHERE enabled = 1 ORDER BY priority ASC, id ASC`

type serverScanner interface {
	Scan(dest ...any) error
}

func scanServer(row *sql.Row) (*server.UsenetServer, error) {
	s, err := scanServerFromRows(row)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, server.ErrNotFound
	}
	return s, err
}

func scanServerFromRows(s serverScanner) (*server.UsenetServer, error) {
	var (
		id            int64
		name          string
		host          string
		port          int
		tls           int
		username      sql.NullString
		password      sql.NullString
		maxConns      int
		priority      int
		enabled       int
		backup        int
		billingMode   string
		quotaBytes    int64
		usedBytes     int64
		bandwidthBPS  int64
		added         int64
		updated       int64
	)
	if err := s.Scan(
		&id, &name, &host, &port, &tls, &username, &password,
		&maxConns, &priority, &enabled,
		&backup, &billingMode, &quotaBytes, &usedBytes, &bandwidthBPS,
		&added, &updated,
	); err != nil {
		return nil, err
	}
	return server.Hydrate(server.HydrateParams{
		ID:                   server.ServerID(id),
		Name:                 name,
		Host:                 host,
		Port:                 port,
		TLS:                  tls != 0,
		Username:             username.String,
		Password:             password.String,
		MaxConns:             maxConns,
		Priority:             priority,
		Enabled:              enabled != 0,
		Backup:               backup != 0,
		BillingMode:          server.BillingMode(billingMode),
		QuotaBytes:           quotaBytes,
		UsedBytes:            usedBytes,
		BandwidthBytesPerSec: bandwidthBPS,
		AddedAt:              time.UnixMilli(added).UTC(),
		UpdatedAt:            time.UnixMilli(updated).UTC(),
	}), nil
}

func boolToInt(b bool) int {
	if b {
		return 1
	}
	return 0
}

func nullableString(s string) any {
	if s == "" {
		return nil
	}
	return s
}

File: src/store/repo_server.zig
//! `servers`: the Usenet provider registry.
//!
//! `ListEnabled` is the orchestrator's dispatch input and runs on every
//! tick, so it rides the partial `servers_active` index over
//! `(priority, id) WHERE enabled = 1` — the index exists precisely so that
//! query never scans disabled rows.
//!
//! `incrementUsedBytes` is a single UPDATE rather than a read-modify-write
//! through `save`, because the byte-accounting flusher fires per download
//! batch and `used_bytes = used_bytes + ?` is atomic against a concurrent
//! flusher by construction. Doing it through the aggregate would be both
//! slower and racy.
//!
//! Passwords are stored in plaintext. Same threat model as SABnzbd and the
//! *arr suite: anybody who can read the data directory can already read
//! everything, and the process needs the cleartext to authenticate against
//! NNTP. Encrypting at rest with a key stored next to the data buys
//! nothing real.

const std = @import("std");
const sqlite = @import("sqlite.zig");
const migrate = @import("migrate.zig");
const server = @import("../domain/server.zig");

const Allocator = std.mem.Allocator;
const Conn = sqlite.Conn;
const UsenetServer = server.UsenetServer;

pub const Error = sqlite.Error || server.RepositoryError || error{
    /// A stored `billing_mode` this binary does not know.
    UnknownBillingMode,
};

const columns =
    "id, name, host, port, tls, username, password, " ++
    "max_conns, priority, enabled, " ++
    "backup, billing_mode, quota_bytes, used_bytes, bandwidth_bytes_per_sec, " ++
    "added_at, updated_at";

/// An owned result set.
pub const ServerList = struct {
    gpa: Allocator,
    items: std.ArrayList(UsenetServer) = .empty,

    pub fn deinit(self: *ServerList) void {
        for (self.items.items) |*s| s.deinit();
        self.items.deinit(self.gpa);
    }
};

pub const ServerRepo = struct {
    conn: *Conn,
    gpa: Allocator,

    pub fn init(gpa: Allocator, conn: *Conn) ServerRepo {
        return .{ .conn = conn, .gpa = gpa };
    }

    pub fn save(self: ServerRepo, s: *UsenetServer) Error!void {
        if (s.id == 0) return self.insert(s);
        return self.update(s);
    }

    fn insert(self: ServerRepo, s: *UsenetServer) Error!void {
        self.conn.execute(
            \\INSERT INTO servers(
            \\    name, host, port, tls, username, password,
            \\    max_conns, priority, enabled,
            \\    backup, billing_mode, quota_bytes, used_bytes, bandwidth_bytes_per_sec,
            \\    added_at, updated_at
            \\) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        , .{
            s.name,
            s.host,
            @as(i64, s.port),
            s.tls,
            // NULL rather than "": some providers permit anonymous access
            // for public groups, and an empty username is a real answer
            // that differs from "no credentials configured".
            sqlite.nullIfEmpty(s.username),
            sqlite.nullIfEmpty(s.password),
            @as(i64, s.max_conns),
            @as(i64, s.priority),
            s.enabled,
            s.backup,
            s.billing_mode.toString(),
            s.quota_bytes,
            s.used_bytes,
            s.bandwidth_bytes_per_sec,
            s.added_at,
            s.updated_at,
        }) catch |e| {
            if (e == error.ConstraintUnique) return error.DuplicateName;
            return e;
        };
        s.setId(self.conn.lastInsertRowid());
    }

    fn update(self: ServerRepo, s: *const UsenetServer) Error!void {
        self.conn.execute(
            \\UPDATE servers SET
            \\    name = ?, host = ?, port = ?, tls = ?, username = ?, password = ?,
            \\    max_conns = ?, priority = ?, enabled = ?,
            \\    backup = ?, billing_mode = ?, quota_bytes = ?, used_bytes = ?,
            \\    bandwidth_bytes_per_sec = ?, updated_at = ?
            \\WHERE id = ?
        , .{
            s.name,
            s.host,
            @as(i64, s.port),
            s.tls,
            sqlite.nullIfEmpty(s.username),
            sqlite.nullIfEmpty(s.password),
            @as(i64, s.max_conns),
            @as(i64, s.priority),
            s.enabled,
            s.backup,
            s.billing_mode.toString(),
            s.quota_bytes,
            s.used_bytes,
            s.bandwidth_bytes_per_sec,
            s.updated_at,
            s.id,
        }) catch |e| {
            if (e == error.ConstraintUnique) return error.DuplicateName;
            return e;
        };
        if (self.conn.changes() == 0) return error.ServerNotFound;
    }

    pub fn byId(self: ServerRepo, gpa: Allocator, id: server.ServerId) Error!UsenetServer {
        return self.one(gpa, "SELECT " ++ columns ++ " FROM servers WHERE id = ?", .{id});
    }

    pub fn byName(self: ServerRepo, gpa: Allocator, name: []const u8) Error!UsenetServer {
        return self.one(gpa, "SELECT " ++ columns ++ " FROM servers WHERE name = ?", .{name});
    }

    /// All servers, priority ascending then id — lower priority number is
    /// higher preference, matching the SABnzbd convention operators
    /// already know.
    pub fn list(self: ServerRepo, gpa: Allocator) Error!ServerList {
        return self.many(gpa, "SELECT " ++ columns ++ " FROM servers ORDER BY priority ASC, id ASC");
    }

    /// The orchestrator's dispatch list.
    pub fn listEnabled(self: ServerRepo, gpa: Allocator) Error!ServerList {
        return self.many(
            gpa,
            "SELECT " ++ columns ++ " FROM servers WHERE enabled = 1 ORDER BY priority ASC, id ASC",
        );
    }

    /// Add `n` to `used_bytes` in one statement.
    ///
    /// Non-positive `n` is a no-op rather than an error: the flusher calls
    /// this unconditionally at the end of a batch, and a batch that
    /// downloaded nothing is normal.
    pub fn incrementUsedBytes(self: ServerRepo, id: server.ServerId, n: i64) Error!void {
        if (n <= 0) return;
        try self.conn.execute(
            "UPDATE servers SET used_bytes = used_bytes + ?, updated_at = ? WHERE id = ?",
            .{ n, sqlite.nowMillis(), id },
        );
        if (self.conn.changes() == 0) return error.ServerNotFound;
    }

    pub fn remove(self: ServerRepo, id: server.ServerId) Error!void {
        try self.conn.execute("DELETE FROM servers WHERE id = ?", .{id});
        if (self.conn.changes() == 0) return error.ServerNotFound;
    }

    fn one(self: ServerRepo, gpa: Allocator, sql: []const u8, args: anytype) Error!UsenetServer {
        var st = self.conn.queryRow(sql, args) catch |e| {
            if (e == error.NoRows) return error.ServerNotFound;
            return e;
        };
        defer st.release();
        return hydrate(gpa, &st);
    }

    fn many(self: ServerRepo, gpa: Allocator, sql: []const u8) Error!ServerList {
        var out = ServerList{ .gpa = gpa };
        errdefer out.deinit();
        var st = try self.conn.query(sql, .{});
        defer st.release();
        while (try st.step()) {
            var s = try hydrate(gpa, &st);
            errdefer s.deinit();
            out.items.append(gpa, s) catch return error.OutOfMemory;
        }
        return out;
    }

    fn hydrate(gpa: Allocator, st: *sqlite.Stmt) Error!UsenetServer {
        return UsenetServer.hydrate(gpa, .{
            .id = st.int(0),
            .name = st.text(1),
            .host = st.text(2),
            .port = @intCast(st.int(3)),
            .tls = st.boolean(4),
            .username = st.text(5),
            .password = st.text(6),
            .max_conns = @intCast(st.int(7)),
            .priority = @intCast(st.int(8)),
            .enabled = st.boolean(9),
            .backup = st.boolean(10),
            .billing_mode = server.BillingMode.parse(st.text(11)) orelse return error.UnknownBillingMode,
            .quota_bytes = st.int(12),
            .used_bytes = st.int(13),
            .bandwidth_bytes_per_sec = st.int(14),
            .added_at = st.int(15),
            .updated_at = st.int(16),
        }) catch return error.OutOfMemory;
    }
};

// ---------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------

const t = std.testing;

fn add(r: ServerRepo, name: []const u8, priority: i32) !UsenetServer {
    var s = try UsenetServer.init(t.allocator, .{
        .name = name,
        .host = "news.example.net",
        .port = 563,
        .username = "u",
        .password = "p",
        .max_conns = 10,
        .priority = priority,
    }, 1);
    errdefer s.deinit();
    try r.save(&s);
    return s;
}

test "save assigns an id and the row round-trips" {
    const conn = try migrate.openMigrated();
    defer conn.close();
    const r = ServerRepo.init(t.allocator, conn);

    var s = try add(r, "main", 0);
    defer s.deinit();
    try t.expect(s.id != 0);

    var got = try r.byId(t.allocator, s.id);
    defer got.deinit();
    try t.expectEqualStrings("main", got.name);
    try t.expectEqualStrings("news.example.net", got.host);
    try t.expectEqual(@as(u16, 563), got.port);
    try t.expectEqual(@as(u16, 10), got.max_conns);
    try t.expectEqualStrings("u", got.username);
    try t.expectEqualStrings("p", got.password);
    try t.expect(got.tls);
    try t.expect(got.enabled);
    try t.expectEqual(server.BillingMode.flat, got.billing_mode);
}

test "byName finds it and a missing name is ServerNotFound" {
    const conn = try migrate.openMigrated();
    defer conn.close();
    const r = ServerRepo.init(t.allocator, conn);
    var s = try add(r, "primary", 0);
    defer s.deinit();

    var got = try r.byName(t.allocator, "primary");
    defer got.deinit();
    try t.expectEqual(s.id, got.id);

    try t.expectError(error.ServerNotFound, r.byName(t.allocator, "missing"));
    try t.expectError(error.ServerNotFound, r.byId(t.allocator, 999));
}

test "list orders by priority ascending" {
    const conn = try migrate.openMigrated();
    defer conn.close();
    const r = ServerRepo.init(t.allocator, conn);
    var backup = try add(r, "backup", 5);
    defer backup.deinit();
    var primary = try add(r, "primary", 0);
    defer primary.deinit();
    var tertiary = try add(r, "tertiary", 10);
    defer tertiary.deinit();

    var got = try r.list(t.allocator);
    defer got.deinit();
    try t.expectEqual(@as(usize, 3), got.items.items.len);
    try t.expectEqualStrings("primary", got.items.items[0].name);
    try t.expectEqualStrings("backup", got.items.items[1].name);
    try t.expectEqualStrings("tertiary", got.items.items[2].name);
}

test "listEnabled skips disabled servers" {
    const conn = try migrate.openMigrated();
    defer conn.close();
    const r = ServerRepo.init(t.allocator, conn);
    var a = try add(r, "a", 0);
    defer a.deinit();
    var b = try add(r, "b", 1);
    defer b.deinit();

    // Disabled, not deleted: the operator keeps the credentials.
    try a.setEnabled(false, 2);
    try r.save(&a);

    var got = try r.listEnabled(t.allocator);
    defer got.deinit();
    try t.expectEqual(@as(usize, 1), got.items.items.len);
    try t.expectEqualStrings("b", got.items.items[0].name);

    // And it is still there in the full list.
    var all = try r.list(t.allocator);
    defer all.deinit();
    try t.expectEqual(@as(usize, 2), all.items.items.len);
}

test "an update rewrites the row" {
    const conn = try migrate.openMigrated();
    defer conn.close();
    const r = ServerRepo.init(t.allocator, conn);
    var s = try add(r, "x", 0);
    defer s.deinit();

    try s.update(.{ .max_conns = 99 }, 5);
    try r.save(&s);

    var got = try r.byId(t.allocator, s.id);
    defer got.deinit();
    try t.expectEqual(@as(u16, 99), got.max_conns);
    try t.expectEqual(@as(i64, 5), got.updated_at);
}

test "remove deletes and is not idempotent" {
    const conn = try migrate.openMigrated();
    defer conn.close();
    const r = ServerRepo.init(t.allocator, conn);
    var s = try add(r, "x", 0);
    defer s.deinit();

    try r.remove(s.id);
    try t.expectError(error.ServerNotFound, r.byId(t.allocator, s.id));
    try t.expectError(error.ServerNotFound, r.remove(s.id));
}

test "a duplicate name is DuplicateName, not a raw constraint error" {
    const conn = try migrate.openMigrated();
    defer conn.close();
    const r = ServerRepo.init(t.allocator, conn);
    var first = try add(r, "dup", 0);
    defer first.deinit();

    var second = try UsenetServer.init(t.allocator, .{
        .name = "dup",
        .host = "h2",
        .port = 119,
        .max_conns = 1,
    }, 1);
    defer second.deinit();
    try t.expectError(error.DuplicateName, r.save(&second));
}

test "incrementUsedBytes accumulates and ignores non-positive deltas" {
    const conn = try migrate.openMigrated();
    defer conn.close();
    const r = ServerRepo.init(t.allocator, conn);
    var s = try add(r, "metered", 0);
    defer s.deinit();

    try r.incrementUsedBytes(s.id, 1000);
    try r.incrementUsedBytes(s.id, 2000);
    // A batch that downloaded nothing is normal, not an error.
    try r.incrementUsedBytes(s.id, 0);
    try r.incrementUsedBytes(s.id, -5);

    var got = try r.byId(t.allocator, s.id);
    defer got.deinit();
    try t.expectEqual(@as(i64, 3000), got.used_bytes);
}

test "incrementUsedBytes on an unknown server is reported" {
    const conn = try migrate.openMigrated();
    defer conn.close();
    const r = ServerRepo.init(t.allocator, conn);
    try t.expectError(error.ServerNotFound, r.incrementUsedBytes(4242, 1));
}

test "a metered backup server round-trips every flag" {
    const conn = try migrate.openMigrated();
    defer conn.close();
    const r = ServerRepo.init(t.allocator, conn);

    var s = try UsenetServer.init(t.allocator, .{
        .name = "block",
        .host = "block.example",
        .port = 563,
        .max_conns = 4,
        .backup = true,
        .billing_mode = .metered,
        .quota_bytes = 500 * 1024 * 1024 * 1024,
        .bandwidth_bytes_per_sec = 10 * 1024 * 1024,
    }, 1);
    defer s.deinit();
    try r.save(&s);

    var got = try r.byId(t.allocator, s.id);
    defer got.deinit();
    try t.expect(got.backup);
    try t.expectEqual(server.BillingMode.metered, got.billing_mode);
    try t.expectEqual(@as(i64, 500 * 1024 * 1024 * 1024), got.quota_bytes);
    try t.expectEqual(@as(i64, 10 * 1024 * 1024), got.bandwidth_bytes_per_sec);
}

test "anonymous credentials come back as empty, and are NULL in the row" {
    const conn = try migrate.openMigrated();
    defer conn.close();
    const r = ServerRepo.init(t.allocator, conn);

    var s = try UsenetServer.init(t.allocator, .{
        .name = "public",
        .host = "free.example",
        .port = 119,
        .max_conns = 1,
        .tls = false,
    }, 1);
    defer s.deinit();
    try r.save(&s);

    try t.expectEqual(@as(i64, 1), try conn.scalarInt(
        "SELECT COUNT(*) FROM servers WHERE username IS NULL AND password IS NULL",
        .{},
    ));
    var got = try r.byId(t.allocator, s.id);
    defer got.deinit();
    try t.expectEqualStrings("", got.username);
    try t.expect(!got.tls);
}

test "updating a row deleted underneath us is reported" {
    const conn = try migrate.openMigrated();
    defer conn.close();
    const r = ServerRepo.init(t.allocator, conn);
    var s = try add(r, "gone", 0);
    defer s.deinit();

    try conn.execute("DELETE FROM servers WHERE id = ?", .{s.id});
    try s.update(.{ .max_conns = 3 }, 9);
    try t.expectError(error.ServerNotFound, r.save(&s));
}

test "an unknown billing mode is rejected rather than guessed" {
    const conn = try migrate.openMigrated();
    defer conn.close();
    const r = ServerRepo.init(t.allocator, conn);
    var s = try add(r, "weird", 0);
    defer s.deinit();

    try conn.exec("PRAGMA ignore_check_constraints = ON");
    try conn.execute("UPDATE servers SET billing_mode = ? WHERE id = ?", .{ "barter", s.id });
    try t.expectError(error.UnknownBillingMode, r.byId(t.allocator, s.id));
}
